pkg/logs/tekton_logging.go

```diff
@@ -247,52 +251,64 @@ func (t TektonLogger) GetRunningBuildLogs(pa *v1.PipelineActivity, buildName str
 	return nil
 }
 
-func (t TektonLogger) getContainerLogsFromPod(pod *corev1.Pod, pa *v1.PipelineActivity, buildName string, stageName string) error {
+func (t *TektonLogger) getContainerLogsFromPod(pod *corev1.Pod, pa *v1.PipelineActivity, buildName string, stageName string) error {
 	infoColor := color.New(color.FgGreen)
 	infoColor.EnableColor()
 	errorColor := color.New(color.FgRed)
 	errorColor.EnableColor()
 	containers, _, _ := kube.GetContainersWithStatusAndIsInit(pod)
-	logsChannel := make(chan LogLine)
-	defer close(logsChannel)
-	errorsChannel := make(chan error)
-	defer close(errorsChannel)
-	go t.LogWriter.StreamLog(logsChannel, errorsChannel)
+	t.initializeLoggingRoutine()
 	for i, ic := range containers {
 		pod, err := t.waitForContainerToStart(pa.Namespace, pod, i, stageName)
 		err = t.LogWriter.WriteLog(LogLine{
-			Line: fmt.Sprintf("\nShowing logs for build %v stage %s and container %s\n",
+			Line: fmt.Sprintf("\nShowing logs for build %v stage %s and container %s",
 				infoColor.Sprintf(buildName), infoColor.Sprintf(stageName), infoColor.Sprintf(ic.Name)),
-		})
+		}, t.logsChannel)
 		if err != nil {
 			return errors.Wrapf(err, "there was a problem writing a single line into the logs writer")
 		}
-		err = t.fetchLogsToChannel(logsChannel, errorsChannel, pa.Namespace, pod, &ic)
+		err = t.fetchLogsToChannel(pa.Namespace, pod, &ic)
 		if err != nil {
 			return errors.Wrap(err, "couldn't fetch logs into the logs channel")
 		}
 		if hasStepFailed(pod, i, t.KubeClient, pa.Namespace) {
-			return t.LogWriter.WriteLog(LogLine{
+			err = t.LogWriter.WriteLog(LogLine{
 				Line: errorColor.Sprintf("\nPipeline failed on stage '%s' : container '%s'. The execution of the pipeline has stopped.", stageName, ic.Name),
-			})
+			}, t.logsChannel)
+			if err != nil {
+				return err
+			}
+			break
 		}
 	}
+	// We are done using the logs and errors channels, any message in the channels should still be read before calling wg.Done()
+	t.closeLoggingChannels()
```

Comment:
so what happens if the code above returns an error? Then we neither sens the closing signal to the channel nor do we wait for the completion of the goroutine. 

This might in most cases not matter, but I think synchronisation is still not correct.

I was also hoping to get the test cases updated to what I had in PR #5194 which made more concise assertions